server/v2/commands.go

```diff
@@ -11,26 +11,49 @@ import (
 
 	"github.com/spf13/cobra"
 
+	"cosmossdk.io/core/transaction"
 	"cosmossdk.io/log"
+	"github.com/spf13/viper"
 )
 
-func Commands(logger log.Logger, homePath string, modules ...ServerModule) (CLIConfig, error) {
+type App[T transaction.Tx] struct {
+	Application Application[T]
+	Store any
+}
+
+type AppCreator[T transaction.Tx] func(*viper.Viper, log.Logger) App[T]
+
+func Commands(rootCmd *cobra.Command, newApp AppCreator[transaction.Tx], logger log.Logger, homePath string, modules ...ServerModule[transaction.Tx]) (CLIConfig, error) {
 	if len(modules) == 0 {
 		// TODO figure if we should define default modules
 		// and if so it should be done here to avoid uncessary dependencies
 		return CLIConfig{}, errors.New("no modules provided")
 	}
 
-	v, err := ReadConfig(filepath.Join(homePath, "config"))
-	if err != nil {
-		return CLIConfig{}, fmt.Errorf("failed to read config: %w", err)
-	}
-
 	server := NewServer(logger, modules...)
+	// Write default config for each server module
+	flags := server.StartFlags()
+
+	if _, err := os.Stat(filepath.Join(homePath, "config", "app.toml")); os.IsNotExist(err) {
```

Comment:
We should not write the config here as we don't have a home flag.
Let's maybe return this whole check as a function, then in AddCommands() we add this in the root command?

```go
func AddCommands(rootCmd *cobra.Command, logger log.Logger, homePath string, modules ...ServerModule) error {
	cmds, configCreator, err := Commands(logger, homePath, modules...)
	if err != nil {
		return err
	}

	rootCmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
		if err := configCreator(cmd); err != nil {
			return err
		}

		if rootCmd.PersistentPreRun != nil {
			rootCmd.PersistentPreRun(cmd, args)
			return nil
		}

		return rootCmd.PersistentPreRunE(cmd, args)
	}

	rootCmd.AddCommand(cmds.Commands...)
	return nil
}
```
Basically wrap the persistent root command that reads the config.
Eventually if this works we can even do more and do the config reading and set the viper in context directly here (so that app devs don't need to do it)

Reply:
It works, makes sense to move both config writing & reading to this path. Wraped into `configHandle` func